### ISSUE TITLE
Add support for timing out slow edits

### ISSUE THREAD

**Issue by depryf** [imsweb/validation#22]
Would apply to running conditions as well. No edits should run forever; but it's very easy to make that happen right now by writing an infinite loop by mistake.

There should be some kind of timeout (it should be a parameter to the framework); the edit should fail if running longer than the timeout.

**Comment by depryf** [comment]
I think to be able to do this, I will need to wrap the executions into a Callable and use an ExecutorServices (probably a fixed threads pool of size 1 since I don't really want to multi-thread this). I can then have a time out on the call itself...

```java
ExecutorService executor = Executors.newCachedThreadPool();
Callable<Object> task = new Callable<Object>() {
   public Object call() {
      return something.blockingMethod();
   }
};
Future<Object> future = executor.submit(task);
try {
   Object result = future.get(5, TimeUnit.SECONDS); 
} catch (TimeoutException ex) {
   // handle the timeout
} catch (InterruptedException e) {
   // handle the interrupts
} catch (ExecutionException e) {
   // handle other exceptions
} finally {
   future.cancel(true); // may or may not desire this
}
```

**Comment by ctmay4** [comment]
I assume that the edit "fails" when it times out.  We just need to make this that is presented to caller.

I can't help but think about whether threading at some point would give a gain.  Not needed for now, but something to consider in the future.  Might not be that hard to implement and it could default to a single thread.

**Comment by depryf** [comment]
Yes, but application dealing with a lot of data are probably already multi-threading the execution of each entities being validated, so I am not sure how much gain we would have. 

Also, the main issue here is that we use Groovy bindings (and edits puts things in that binding, they don't just read it), and the binding itself is not thread-safe. It's not a very simple problem to solve (I have looked into it before).

I agree it's a neat idea, but for now I will start with one thread so I don't run into those racing conditions.

**Comment by depryf** [comment]
This is done (disabled by default). To enable this new feature, call 
```java
ValidationEngine.enableEditExecutionTimeout(n);
```
where n is the number of second after which a slow edit should fail.

The same mechanism is used for conditions.

This will be available with the release of version 1.7.
